# GoodMethod
short description of GoodMethod

## Requirement
  * Linux/Unix environment 
  * Python
  * beanplot (R packages)
  * samtools
  * bedtools
  * (optional) Rphylip with Phylip for clustering and generating phylogenetic tree. 

## Input
  1.  tab-delimited __bed__ file of CNV segments information from exom sequencing.    
    ```
   [chromosome]	[start]	[end]	[chromosome:start:end:CNV]
   ```
    __Note: the 4th column of the file must have the exact same format with that of below.__ (__Amp__:amplification, __Del__:deletion)

  example
	```
	7   19533   157408385   7:19533:157408385:Amp
	9   19116859    32405639    9:19116859:32405639:Del
	```
	
  2.  alignment files (.bam) for single-cell RNA-seq of __tumor__ sample
  3.  alignment files (.bam) for single-cell RNA-seq of __normal__ control

## Config file
Adjusting __GoodMethod.cfg__ can change the belows.
  * Path to python/samtools/bedtools/Rscript
  * The threshold of mapping-qualities/read-count for alignment files
  * FDR for CNV calling


## Running
  ```
  cd ./GoodMethod/
  bash run_GoodMethod.sh [directory_for_tumor] [directory_for_normal] [tab-delimited bed file for CNV segments information] [base name for output file]
  ```
  * __[directory for tumor]__: path to directory which aligned bam files of __tumor__ single cells are in.
  * __[directory for normal]__: path to directory which aligned bam files of __normal__ control single cells are in.  

## Output
The directory __output_[base name]__ would be generated and all the output files would be located in this directory.
  1. __incidenceMatrix.csv__: the matrix of the presence/absence of the CNVs in individual single cells
  2. __pdf__: 
    1. Read-count distribution in CNV segments. (violin plot)
    2. Hierarchical clustering of the tumor single cells by the presence/absence of the CNVs.

## (optional) Constructing phylogenetic tree
Phylogenetic trees could be generated by the presence/absence profiles of CNV. This requires additional R package (Rphylip) for execution.
  * Before running, the path to phylip need to be added in __Tree.cfg__ file.
```
cd ./GoodMethod/
bash run_Tree.sh [CNV presence/absence matrix][number of genotypes] [base name for output file]
```
  * __[CNV presence/absence matrix]__: .incidenceMatrix.csv files. Note that exclusion of non-malignent single cells from the matrix is recommended. 
  * __[number of genotypes]__: the number of genotypes which are used for constructing trees by the aid of hierarchical clustering from the previous output files.  




